### DISCUSSION TITLE
Some Questions on Coupling SPPARKS and MyTrim Modules in Magpie

### DISCUSSION THREAD

**Discussion by mcrossover97** [idaholab/moose#18986] [Q&A General] [answered]
Hi there,

I am trying to understand how exactly the coupling of SPPARKS and MyTrim with Magpie works. I have several questions:

1. When compiling SPPARKS and MyTrim, a bunch of `.lo` and `.d` files are created. I assume that they are object files that will be used to make `libmagpie-opt` files in the `lib` folder. Am I correct? If so, what is the difference between `.lo` and `.d` files? One can understand from `spparks.mk` and `mytrim.mk` files that `.lo`  and `.d` files are labeled as `ADDITIONAL_APP_OBJECTS` and `ADDITIONAL_APP_DEPS`, respectively. However, it seems that when I delete the following line which is used to make `.d` files, `.d` files will not be created and yet no error comes up during compilation:

`ADDITIONAL_APP_DEPS += $(patsubst %.C, %.$(obj-suffix).d, $(mytrim_srcfiles))`

What are they for and are they even necessary?

2. From my understanding, when compiling a module, all `.c` and `.cpp` files used in that module should be compiled. However, for example, in the MyTrim module, only files that are in the module main folder (`MYTRIM_DIR`) are made into object files while there is another `.c` files in the `$(MYTRIM_DIR)/shim` folder (`point.c`) that seems to go uncompiled. Anyway, it looks like files that are in this folder (`point.c`, `point.h`, and `pow.h`) aren't doing anything. This is because, when I delete the codes within these 3 files, the program will be compiled without error although `simconf.h` in `$(MYTRIM_DIR)` seems to be using the `point` class. What am I missing here?

3.  I don't have GSL and fftw3 installed. Shouldn't it cause an error when compiling? Where GSL and fftw3 are used?

4. What does `ADDITIONAL_CPPFLAGS += -DMYTRIM_ENABLED` in the `magpie.mk` file do? If I delete this line, I encounter a compilation error.

5. It seems like MyTrim uses CMake to generate a makefile as we can see the `CMakeLists.txt` in the main folder. However, I think that this file doesn't impact compilation. How does our compiler understand what procedures to follow to compile a program without using CMake to generate a makefile? For example, if a package uses C++17 features, I think we can set this in the CMake file which in turn generates a Makefile that is told to compile based on C++17. How do we set stuff like this in our makefiles without CMake?

I know probably my questions are full of insufficient, wrong, and misguiding terminologies due to my lack of knowledge in computer engineering. I hope I have been able to express what I mean. I would be really appreciative if someone could clear things up for me. Thanks for your help and support,
Mohammad

**Comment by GiudGiud** [comment]
@cticenhour who is in charge of Magpie?

**Reply by cticenhour** [reply]
@dschwen are you still in charge of Magpie?

**Reply by GiudGiud** [reply]
@snschune @recuero

**Reply by dschwen** [reply]
Me

**Comment by dschwen** [comment, marked as answer]
1 .`.d` files are dependency files. We use the compiler supplied dependency resolution to enable recompiling only what needs to be recompiled.
Our build system makes heavy use of libtool; `.lo` files contain information for building a `.so`.
2. You are missing that myTRIM is a submodule in Magpie, and it is also intended to be compiled as a standalone program. The files you mention contain shims for objects that MOOSE provides, but are absent (and need to come from somewhere) when building myTRIM as a standalone code.
3. GSL and FFTW3 are optional dependencies. If you look closer in the Makefiles you'll see `-D` defines that are checked using `#ifdef`s which disable certain classes in Magpie if those libs are not available
4. what I just said ^
5. See my answer to 2. CMake is only used when building a standalone version of myTRIM
